refactored dashboard_2 to use apis

aryanbhosale opened the pull request:
# Pull Request

## Description

created apis for 4 options - no inverter, solis, givenergy, and enphase

https://github.com/user-attachments/assets/a7b1d3ac-b2c0-4c0d-bf23-45ebbbda9b88



## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

Sukh-P commented:
@aryanbhosale having a look at this again I think that it would be worth taking the time to refactor the enphase functions that already exist so they can be reused in the API, it feels like there is too much enphase specific logic in that file currently to justify since we would want to keep that file as clean as possible for people to expand it and understand it as easily as possible. 

@peterdudfield might be worth you taking a look at this PR too just to make sure you're happy with the direction of the API changes

aryanbhosale commented:
> @aryanbhosale having a look at this again I think that it would be worth taking the time to refactor the enphase functions that already exist so they can be reused in the API, it feels like there is too much enphase specific logic in that file currently to justify since we would want to keep that file quite clean for people to expand it and understand it as easily as possible.
> 
> @peterdudfield might be worth you taking a look at this PR too just to make sure you're happy with the direction of the API changes

Yeah I will try to reuse the enphase functions as much as possible, but we're constrained here due to the fact that most of the enphase functions are CLI-specific, ill try to do it for the ones that aren't, thank you for the review
Are you happy with the direction the api is going in?

Sukh-P commented on api/app/api.py:
```diff
@@ -19,7 +41,105 @@
+        solis_data = None
+        givenergy_data = None
+        
+        if site.inverter_type:
```
I am not quite sure about this part here, we are getting the inverter data here but they can also be got in make_pv_data later?

aryanbhosale replied:
> I am not quite sure about this part here, we are getting the inverter data here but they can also be got in make_pv_data later?

this is the prefetch code
so to break this down, tmake_pv_data can fetch data when called from cli and no params(see optional params of that func) are provided - this is in the CLI context
but when the data is provided as params(prefetched data), it can be used in the API context beofr even reaching the CLI part

peterdudfield replied:
hmm, i think i agree with @Sukh-P, would it not be simplier if it all goes in make_pv_data later

aryanbhosale replied:
> hmm, i think i agree with @Sukh-P, would it not be simplier if it all goes in `make_pv_data` later

it woudlve made sense if some of the functions didnt ask for an auth url in the CLI- which we cant afford to have in the website

peterdudfield replied:
see other comments, I think we are talking about all the same thing

Sukh-P commented on quartz_solar_forecast/data.py:
```diff
@@ -179,33 +179,48 @@ def process_pv_data(live_generation_kw: Optional[pd.DataFrame], ts: pd.Timestamp
     return da
 
-def make_pv_data(site: PVSite, ts: pd.Timestamp) -> xr.Dataset:
+def make_pv_data(
```
Linked to my previous comment not sure about having all of these as input parameters vs some of them being created in the function like `givenergy_data`

aryanbhosale replied:
> Linked to my previous comment not sure about having all of these as input parameters vs some of them being created in the function like `givenergy_data`

thats the reason we're first checking in every if inverter_type=="xyz" condition whether the data already exists or not, that already existing data is the prefetched one from the api when the api is used, but when running from the cli, that data would be None, so we run the func to fetch it from the respective inverter file

peterdudfield replied:
would it be better to just set the enphase token, and then you can sue `make_pv_data` to get the data. You might need some documentation, or some direction how people get that token

aryanbhosale replied:
> would it be better to just set the enphase token, and then you can sue `make_pv_data` to get the data. You might need some documentation, or some direction how people get that token

the enphase token cant be set, it has an additional level of authentication wwhich requires the home owner to approve - which will give the user the access token on demand

peterdudfield replied:
Ok, im just looking here

https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/inverters/enphase.py#L52

Currently 
1. get enphase auth url
2. get_enphase_authorization_code (which currently uses CLI)
3. get access token using the `auth code`

Couldnt there be a slightly seperate process to get the `auth code`, then you just set the `auth code` in your env vars and you can then automatically get the access token

aryanbhosale replied:
> Ok, im just looking here
> 
> https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/inverters/enphase.py#L52
> 
> Currently
> 
> 1. get enphase auth url
> 2. get_enphase_authorization_code (which currently uses CLI)
> 3. get access token using the `auth code`
> 
> Couldnt there be a slightly seperate process to get the `auth code`, then you just set the `auth code` in your env vars and you can then automatically get the access token

the thing with auth code isthat it keeps changing , and it's like an OTP, so we cant reuse it

peterdudfield replied:
How many times can you use it? 

Could you then going abd get the access token, in a different script. 

Then just set the access token in an env var ready to use

aryanbhosale replied:
> How many times can you use it?
> 
> Could you then going abd get the access token, in a different script.
> 
> Then just set the access token in an env var ready to use

we can look at it this way, everytime you log on to the website, and want the enphase data, you need that auth code(which can only be used once) to get the access token, the access token has some validity time, which im storing in a session state, so we can keep usig that in the website

aryanbhosale replied:
> How many times can you use it?
> 
> Could you then going abd get the access token, in a different script.
> 
> Then just set the access token in an env var ready to use

and also this is what im doing in the website, just instead of an env var,im using session states,since env var would mean a more permanent solution(which is pointless since access tokens do expire)

peterdudfield replied:
How long do they take to expire?

aryanbhosale replied:
> How long do they take to expire?

not quite sure, but if youre not a system owner then 1day if i rem correctly

Sukh-P commented:
@aryanbhosale thanks for making these changes! I think we are getting there but still need to think about the structure of the code a bit to get the abstractions right, one thing is that the API would become significantly more complex now so I think adding some documentation/examples in this PR would be very helpful

peterdudfield commented on api/app/api.py:
```diff
@@ -19,7 +41,105 @@
     allow_headers=["*"]
 )
 
+class ForecastRequest(BaseModel):
```
Could you use this to a models.py file in the same dir

aryanbhosale replied:
> Could you use this to a models.py file in the same dir

yes sure

aryanbhosale replied:
> > Could you use this to a models.py file in the same dir
> 
> yes sure

done this in my recent commit

peterdudfield commented on api/app/api.py:
```diff
@@ -19,7 +41,105 @@
+    access_token = request.access_token
+    enphase_system_id = request.enphase_system_id
+
+    try:
```
please remove this try and except, fast api autoamtically gives a 500 if something goes wrong, so I dont think this is needed

aryanbhosale replied:
> please remove this try and except, fast api autoamtically gives a 500 if something goes wrong, so I dont think this is needed

alright will do

aryanbhosale replied:
> please remove this try and except, fast api autoamtically gives a 500 if something goes wrong, so I dont think this is needed

removed this try catch block

peterdudfield commented on quartz_solar_forecast/data.py:
```diff
@@ -179,33 +179,48 @@ def process_pv_data(live_generation_kw: Optional[pd.DataFrame], ts: pd.Timestamp
-        try:
-            live_generation_kw = get_givenergy_data()
-        except Exception as e:
-            print(f"Error retrieving GivEnergy data: {str(e)}")
```
i probably prefer it like it was, rather than pasing in the data. Perhaps ive missed a reason why you changed this?

aryanbhosale replied:
> i probably prefer it like it was, rather than pasing in the data. Perhaps ive missed a reason why you changed this?

since im reusing this function(make_pv_data) in the API context, and given the fact that this function is linked to get_enphase_data(), which only supports CLI inputs for the auth url and so on, we cant rely on entering our auth url in the CLI when working on the web, so to make everything streamlined, ive kept those solis/givenergy/enphase data as optional params to this function, so that IF theyre passed from the API context, they wont reach the CLI part and the fetching happens smoothly all on the web
but when those Optional params of inverter data arent provided, theyre None by default, then we rely on those CLI based functions to fetch the inverter data

peterdudfield replied:
could you refactro get_enphase_data() to work of the token? and then you just have a different way to get the token

peterdudfield commented:
I decided to draw a diagram, as I think this might help the discussion. 

This is what I think the UI, API interaction should be

With no inverters

<img width="483" alt="Screenshot 2024-08-05 at 18 58 18" src="https://github.com/user-attachments/assets/a83e584b-81b5-45a3-9ba6-c9ca91ed7db2">

With inverters

<img width="561" alt="Screenshot 2024-08-05 at 18 58 59" src="https://github.com/user-attachments/assets/60625b76-6d42-48a3-afd5-aa69ed26932f">

Do you think its a good idea to change the code to this?

aryanbhosale commented:
> I decided to draw a diagram, as I think this might help the discussion.
> 
> This is what I think the UI, API interaction should be
> 
> With no inverters
> 
> <img alt="Screenshot 2024-08-05 at 18 58 18" width="483" src="https://private-user-images.githubusercontent.com/34686298/355160898-a83e584b-81b5-45a3-9ba6-c9ca91ed7db2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA4OTgtYTgzZTU4NGItODFiNS00NWEzLTliYTYtYzljYTkxZWQ3ZGIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNDkwODhiMWM4YWUyOGExZmRhNWM4MDg5YmJjOTEyYzNiMzc2MjJkOWZjYTllOGU1NWY3MTcxZWZkN2VjZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3TKv-CBYyjAGss5FeQt5YmD2DcIAaG6h6IM3BlODbuM">
> With inverters
> 
> <img alt="Screenshot 2024-08-05 at 18 58 59" width="561" src="https://private-user-images.githubusercontent.com/34686298/355160926-60625b76-6d42-48a3-afd5-aa69ed26932f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA5MjYtNjA2MjViNzYtNmQ0Mi00OGEzLWFmZDUtYWE2OWVkMjY5MzJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OGQ1YjIzZWU5NTc2ODc4MWEwMjc1YWU5ODNiZDJmN2I2YzY2MTI5ZDcxN2M3ZjZkZGIwOGU1YWJiYjk5YjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PiwmWQ8t9sEQt_EvvuxPGxzpZk_jyc5kvXeCZ9_xsq0">
> Do you think its a good idea to change the code to this?

Thank you @peterdudfield for making these comprehensive diagrams, the one without any inverter type looks good. 
There are slight intricacies in the one with inverters specified:
1.The part till "Get forecast" to "run forecast" is fine
2. Run forecast has dependency on make pv data, which has another dependency on get enphase data (only for enphase inverter) - now this get enphase data function has certain functions that ask for auth url and so on in the CLI, which cant be reused and will have to be rewritten for the API purpose like it is happening currently in the main code(not this pr)

aryanbhosale commented:
> > I decided to draw a diagram, as I think this might help the discussion.
> > This is what I think the UI, API interaction should be
> > With no inverters
> > <img alt="Screenshot 2024-08-05 at 18 58 18" width="483" src="https://private-user-images.githubusercontent.com/34686298/355160898-a83e584b-81b5-45a3-9ba6-c9ca91ed7db2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA4OTgtYTgzZTU4NGItODFiNS00NWEzLTliYTYtYzljYTkxZWQ3ZGIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNDkwODhiMWM4YWUyOGExZmRhNWM4MDg5YmJjOTEyYzNiMzc2MjJkOWZjYTllOGU1NWY3MTcxZWZkN2VjZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3TKv-CBYyjAGss5FeQt5YmD2DcIAaG6h6IM3BlODbuM">
> > With inverters
> > <img alt="Screenshot 2024-08-05 at 18 58 59" width="561" src="https://private-user-images.githubusercontent.com/34686298/355160926-60625b76-6d42-48a3-afd5-aa69ed26932f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA5MjYtNjA2MjViNzYtNmQ0Mi00OGEzLWFmZDUtYWE2OWVkMjY5MzJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OGQ1YjIzZWU5NTc2ODc4MWEwMjc1YWU5ODNiZDJmN2I2YzY2MTI5ZDcxN2M3ZjZkZGIwOGU1YWJiYjk5YjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PiwmWQ8t9sEQt_EvvuxPGxzpZk_jyc5kvXeCZ9_xsq0">
> > Do you think its a good idea to change the code to this?
> 
> Thank you @peterdudfield for making these comprehensive diagrams, the one without any inverter type looks good. There are slight intricacies in the one with inverters specified: 1.The part till "Get forecast" to "run forecast" is fine 2. Run forecast has dependency on make pv data, which has another dependency on get enphase data (only for enphase inverter) - now this get enphase data function has certain functions that ask for auth url and so on in the CLI, which cant be reused and will have to be rewritten for the API purpose like it is happening currently in the main code(not this pr)

to tackle this very issue without having to rewrite anything again, and with every function being reused, i changed the signature(but optionally), to make it work for both the cases, so the functions that ive changed will work just fine normally too, but will work with both CLI and the api

peterdudfield commented:
> > > I decided to draw a diagram, as I think this might help the discussion.
> > > This is what I think the UI, API interaction should be
> > > With no inverters
> > > <img alt="Screenshot 2024-08-05 at 18 58 18" width="483" src="https://private-user-images.githubusercontent.com/34686298/355160898-a83e584b-81b5-45a3-9ba6-c9ca91ed7db2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA4OTgtYTgzZTU4NGItODFiNS00NWEzLTliYTYtYzljYTkxZWQ3ZGIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNDkwODhiMWM4YWUyOGExZmRhNWM4MDg5YmJjOTEyYzNiMzc2MjJkOWZjYTllOGU1NWY3MTcxZWZkN2VjZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3TKv-CBYyjAGss5FeQt5YmD2DcIAaG6h6IM3BlODbuM">
> > > With inverters
> > > <img alt="Screenshot 2024-08-05 at 18 58 59" width="561" src="https://private-user-images.githubusercontent.com/34686298/355160926-60625b76-6d42-48a3-afd5-aa69ed26932f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA5MjYtNjA2MjViNzYtNmQ0Mi00OGEzLWFmZDUtYWE2OWVkMjY5MzJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OGQ1YjIzZWU5NTc2ODc4MWEwMjc1YWU5ODNiZDJmN2I2YzY2MTI5ZDcxN2M3ZjZkZGIwOGU1YWJiYjk5YjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PiwmWQ8t9sEQt_EvvuxPGxzpZk_jyc5kvXeCZ9_xsq0">
> > > Do you think its a good idea to change the code to this?
> > 
> > 
> > Thank you @peterdudfield for making these comprehensive diagrams, the one without any inverter type looks good. There are slight intricacies in the one with inverters specified: 1.The part till "Get forecast" to "run forecast" is fine 2. Run forecast has dependency on make pv data, which has another dependency on get enphase data (only for enphase inverter) - now this get enphase data function has certain functions that ask for auth url and so on in the CLI, which cant be reused and will have to be rewritten for the API purpose like it is happening currently in the main code(not this pr)
> 
> to tackle this very issue without having to rewrite anything again, and with every function being reused, i changed the signature(but optionally), to make it work for both the cases, so the functions that ive changed will work just fine normally too, but will work with both CLI and the api

Does your recent merge change this? 

Could you refactor `get enphase data` to not have the CLI part? Perhaps this could be done in a different PR first?

aryanbhosale commented:
> > > > I decided to draw a diagram, as I think this might help the discussion.
> > > > This is what I think the UI, API interaction should be
> > > > With no inverters
> > > > <img alt="Screenshot 2024-08-05 at 18 58 18" width="483" src="https://private-user-images.githubusercontent.com/34686298/355160898-a83e584b-81b5-45a3-9ba6-c9ca91ed7db2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA4OTgtYTgzZTU4NGItODFiNS00NWEzLTliYTYtYzljYTkxZWQ3ZGIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNDkwODhiMWM4YWUyOGExZmRhNWM4MDg5YmJjOTEyYzNiMzc2MjJkOWZjYTllOGU1NWY3MTcxZWZkN2VjZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3TKv-CBYyjAGss5FeQt5YmD2DcIAaG6h6IM3BlODbuM">
> > > > With inverters
> > > > <img alt="Screenshot 2024-08-05 at 18 58 59" width="561" src="https://private-user-images.githubusercontent.com/34686298/355160926-60625b76-6d42-48a3-afd5-aa69ed26932f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI4NzI2MjYsIm5iZiI6MTcyMjg3MjMyNiwicGF0aCI6Ii8zNDY4NjI5OC8zNTUxNjA5MjYtNjA2MjViNzYtNmQ0Mi00OGEzLWFmZDUtYWE2OWVkMjY5MzJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA1VDE1Mzg0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OGQ1YjIzZWU5NTc2ODc4MWEwMjc1YWU5ODNiZDJmN2I2YzY2MTI5ZDcxN2M3ZjZkZGIwOGU1YWJiYjk5YjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PiwmWQ8t9sEQt_EvvuxPGxzpZk_jyc5kvXeCZ9_xsq0">
> > > > Do you think its a good idea to change the code to this?
> > > 
> > > 
> > > Thank you @peterdudfield for making these comprehensive diagrams, the one without any inverter type looks good. There are slight intricacies in the one with inverters specified: 1.The part till "Get forecast" to "run forecast" is fine 2. Run forecast has dependency on make pv data, which has another dependency on get enphase data (only for enphase inverter) - now this get enphase data function has certain functions that ask for auth url and so on in the CLI, which cant be reused and will have to be rewritten for the API purpose like it is happening currently in the main code(not this pr)
> > 
> > 
> > to tackle this very issue without having to rewrite anything again, and with every function being reused, i changed the signature(but optionally), to make it work for both the cases, so the functions that ive changed will work just fine normally too, but will work with both CLI and the api
> 
> Does your recent merge change this?
> 
> Could you refactor `get enphase data` to not have the CLI part? Perhaps this could be done in a different PR first?

My recent commit does change this, but my enphase code doesn't work after that, and yeah perhaps I could make the CLI part work in a different code, anyway it's just for testing and for the people who know how to code, I suppose the api part would be more important because more people would be able to use it

Sukh-P commented on api/app/models.py:
```diff
@@ -0,0 +1,13 @@
+from typing import Optional
```
Do you think naming this file request_models or schemas might be a little clearer @aryanbhosale?

aryanbhosale replied:
Yeah even I feel so , schemas sounds better

aryanbhosale commented on api/app/api.py:
```diff
@@ -20,6 +41,73 @@
+    
+    return response
+
+@app.get("/enphase/auth_url")
```
Yup can be done

Sukh-P commented:
So I think this already looks much cleaner @aryanbhosale, nice work on that! I have left a few comments, I think it may be nice to get rid of the CLI part in get_enphase_data since that seems like the main blocker in making this even more streamlined code wise, but I would not be opposed to this being merged in first and then doing that straight afterwards. @peterdudfield what do you feel about that? Happy for that change to come first though because it would for me make this PR even better then

aryanbhosale commented:
> So I think this already looks much cleaner @aryanbhosale, nice work on that! I have left a few comments, I think it may be nice to get rid of the CLI part in get_enphase_data since that seems like the main blocker in making this even more streamlined code wise, but I would not be opposed to this being merged in first and then doing that straight afterwards. @peterdudfield what do you feel about that? Happy for that change to come first though because it would for me make this PR even better then
> 
> Also still have some merge conflicts that need to be resolved since adding in the Solarman integration I think

Thank you Sukh, yes even I feel we could remove the CLI part from the enphase code, and yea I could try pushing that change before merging this in.

peterdudfield commented:
> So I think this already looks much cleaner @aryanbhosale, nice work on that! I have left a few comments, I think it may be nice to get rid of the CLI part in get_enphase_data since that seems like the main blocker in making this even more streamlined code wise, but I would not be opposed to this being merged in first and then doing that straight afterwards. @peterdudfield what do you feel about that? Happy for that change to come first though because it would for me make this PR even better then
> 
> Also still have some merge conflicts that need to be resolved since adding in the Solarman integration I think

Fine from me, to merge in which order you want. Good to make sure all tests are passing when merging to main though

aryanbhosale commented:
> > So I think this already looks much cleaner @aryanbhosale, nice work on that! I have left a few comments, I think it may be nice to get rid of the CLI part in get_enphase_data since that seems like the main blocker in making this even more streamlined code wise, but I would not be opposed to this being merged in first and then doing that straight afterwards. @peterdudfield what do you feel about that? Happy for that change to come first though because it would for me make this PR even better then
> > Also still have some merge conflicts that need to be resolved since adding in the Solarman integration I think
> 
> Fine from me, to merge in which order you want. Good to make sure all tests are passing when merging to main though

The enphase api part doesnt work, so ill fix that first then we can get this merged in

Sukh-P commented on api/app/utils/enphase_utils.py:
```diff
@@ -0,0 +1,131 @@
+from quartz_solar_forecast.data import get_nwp, process_pv_data
+from quartz_solar_forecast.inverters.enphase import process_enphase_data
+
+def get_enphase_auth_url():
```
This is duplicated now I think in this file and the original enphase file? Would be good to double check if other are also duplicated and remove duplications were possible

aryanbhosale replied:
> This is duplicated now I think in this file and the original enphase file? Would be good to double check if other are also duplicated and remove duplications were possible

Thanks for pointing thids out, ive removed it in my recent commit

Sukh-P commented on api/app/api.py:
```diff
@@ -20,6 +30,62 @@
+    site_no_live = PVSite(latitude=site.latitude, longitude=site.longitude, capacity_kwp=site.capacity_kwp)
+    predictions_no_live = run_forecast(site=site_no_live, ts=timestamp, nwp_source=nwp_source)
+
+    if site.inverter_type == 'enphase' and access_token and enphase_system_id:
```
I think the first check should potentially be if there is an inverter type at all, if it's None you can just run the regular forecast and just return that no need for the live and no live columns

aryanbhosale replied:
> I think the first check should potentially be if there is an inverter type at all, if it's None you can just run the regular forecast and just return that no need for the live and no live columns

Don't we know that from the checks? No Inverter is also a type of inverter, but it doesn't exist in the pvsite class , so I treats it as no inverter

Sukh-P replied:
I would look at the else statement you have at the moment again, it duplicates data at the moment for the case with no inverter which will most likely be the most common type of use of this API I imagine

aryanbhosale replied:
> I would look at the else statement you have at the moment again, it duplicates data at the moment for the case with no inverter which will most likely be the most common type of use of this API I imagine

Do I keep a separate else condition for no inverter?

aryanbhosale replied:
> I would look at the else statement you have at the moment again, it duplicates data at the moment for the case with no inverter which will most likely be the most common type of use of this API I imagine

ive made this change @Sukh-P

Sukh-P commented on quartz_solar_forecast/inverters/solarman.py:
```diff
@@ -76,4 +76,4 @@ def get_solarman_data(start_date, end_date):
     full_data = full_data.sort_values('timestamp')
     
-    return full_data
+    return full_data    
```
Is this added space an accident?

aryanbhosale replied:
> Is this added space an accident?

Yes my bad

Sukh-P commented:
Nice work @aryanbhosale I think this is looking much better now, one thing still missing is some documentation for the API, since it is becoming more complex here it would be great if you could add that in as well

aryanbhosale commented:
> Nice work @aryanbhosale I think this is looking much better now, one thing still missing is some documentation for the API, since it is becoming more complex here it would be great if you could add that in as well

I have added a readme for the new api , what do I need to add there?

peterdudfield commented on api/app/utils/enphase_utils.py:
```diff
@@ -0,0 +1,126 @@
+    da = process_pv_data(live_generation_kw, ts, site)
+    return da
+
+def predict_ocf(
```
im suprised this function is in there? Is this duplicated?

aryanbhosale replied:
> im suprised this function is in there? Is this duplicated?

Since I'm passing the access token for it to persist in the app, it needed to be modified, it's how it's done in the current main branch too

peterdudfield replied:
I understand you need to pass through the access token. I think there is a neater way of doing this, rather than duplicating the code.

You could
1. Pass through all relevant access tokens, ids e.t.c in a dict in the `ForecastRequest` when calling the api. This could be an environmental variables dict.
2. Set these env vars as necessary for the inverters
3. Call 'run_forecast` from the main library. 

We might need to think slightly carefully how we set the api and dashbaord up, as I think sending secrets thorugh http is considered unsafe. Perhaps we could encode them somehow if necessary, but this can be later.

aryanbhosale replied:
> I understand you need to pass through the access token. I think there is a neater way of doing this, rather than duplicating the code.
> 
> You could
> 
> 1. Pass through all relevant access tokens, ids e.t.c in a dict in the `ForecastRequest` when calling the api. This could be an environmental variables dict.
> 2. Set these env vars as necessary for the inverters
> 3. Call 'run_forecast` from the main library.
> 
> We might need to think slightly carefully how we set the api and dashbaord up, as I think sending secrets thorugh http is considered unsafe. Perhaps we could encode them somehow if necessary, but this can be later.
> 
> Doing it this way thought, should reduce the duplication of code massively

This sounds promising, we could try hashing the token perhaps while passing, but this can be thought of later after getting it working with env vars, thanks for the direction will try this out

aryanbhosale commented on api/app/utils/enphase_utils.py:
```diff
@@ -0,0 +1,126 @@
+    solarman_data: pd.DataFrame = None
+) -> pd.DataFrame:
+    if model == "gb":
+        return predict_ocf(site, None, ts, nwp_source, access_token, enphase_system_id, solis_data, givenergy_data, solarman_data)
```
> should these all be in here?

It's needed because their signatures are different

Sukh-P commented:
> > Nice work @aryanbhosale I think this is looking much better now, one thing still missing is some documentation for the API, since it is becoming more complex here it would be great if you could add that in as well
> 
> I have added a readme for the new api , what do I need to add there?

I can't see that README in this PR? For the API or the streamlitt app?

aryanbhosale commented:
> > > Nice work @aryanbhosale I think this is looking much better now, one thing still missing is some documentation for the API, since it is becoming more complex here it would be great if you could add that in as well
> > 
> > 
> > I have added a readme for the new api , what do I need to add there?
> 
> I can't see that README in this PR? For the API or the streamlitt app?

The streamlit app within dashboard

Sukh-P commented:
> > > > Nice work @aryanbhosale I think this is looking much better now, one thing still missing is some documentation for the API, since it is becoming more complex here it would be great if you could add that in as well
> > > 
> > > 
> > > I have added a readme for the new api , what do I need to add there?
> > 
> > 
> > I can't see that README in this PR? For the API or the streamlitt app?
> 
> The streamlit app within dashboard

I think the API itself could do with some documentation there is none currently and as I mentioned by adding more endpoints to it it is getting more complex, so some docs on even basic ones explaining the endpoints you have added and what is returned and the parameters you can pass would be good

aryanbhosale commented:
> > > > > Nice work @aryanbhosale I think this is looking much better now, one thing still missing is some documentation for the API, since it is becoming more complex here it would be great if you could add that in as well
> > > > 
> > > > 
> > > > I have added a readme for the new api , what do I need to add there?
> > > 
> > > 
> > > I can't see that README in this PR? For the API or the streamlitt app?
> > 
> > 
> > The streamlit app within dashboard
> 
> I think the API itself could do with some documentation there is none currently and as I mentioned by adding more endpoints to it it is getting more complex, so some docs on even basic ones explaining the endpoints you have added and what is returned and the parameters you can pass would be good

Ah alright that way, will do

aryanbhosale commented:
> docs on even basic ones explaining the endpoints you have added and what is returned and the parameters you can pass would be good

added these in my recent commit

peterdudfield commented on api/app/api.py:
```diff
@@ -20,6 +30,65 @@
-    df =run_forecast(site)
-    return df.to_dict()
\ No newline at end of file
+def forecast(request: ForecastRequest):
```
could you rename request to forecast_request, request sometimes has a different meaning in apis (it can be the http request)

aryanbhosale replied:
> could you rename request to forecast_request, request sometimes has a different meaning in apis (it can be the http request)

Sure can do that

peterdudfield replied:
Thank you

peterdudfield commented on api/app/utils/enphase_utils.py:
```diff
@@ -0,0 +1,126 @@
+from quartz_solar_forecast.data import get_nwp, process_pv_data
+from quartz_solar_forecast.inverters.enphase import process_enphase_data
+
+def get_enphase_access_token(auth_code):
```
this feels duplicate of https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/inverters/enphase.py#L52

peterdudfield commented on api/app/utils/enphase_utils.py:
```diff
@@ -0,0 +1,126 @@
+
+    return data_json["access_token"]
+
+def get_enphase_data(enphase_system_id: str, access_token: str) -> pd.DataFrame:
```
this might also be duplicated

peterdudfield commented on api/app/utils/enphase_utils.py:
```diff
@@ -0,0 +1,126 @@
+    data_json = json.loads(data.decode("utf-8"))
+    return process_enphase_data(data_json, start_at)
+
+def make_pv_data(
```
im confused here, why dont you use? https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/data.py#L234

aryanbhosale replied:
> im confused here, why dont you use? https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/data.py#L234

I'm passing the access token through run_forecast which uses make pv data and passing this token to every enphase request, but I think by using env vars as you suggested , this could be overcome

peterdudfield replied:
did you do this with env vars?

aryanbhosale replied:
> did you do this with env vars?

Yes, within the enphase.py file, when an auth code is sent from the streamlit Frontend, it stores the access and refresh tokens in env vars

peterdudfield commented on api/app/api.py:
```diff
@@ -1,16 +1,26 @@
+from dotenv import load_dotenv
 from quartz_solar_forecast.forecast import run_forecast
+from quartz_solar_forecast.pydantic_models import PVSite, ForecastRequest, TokenRequest
+from .utils.enphase_utils import get_enphase_access_token, get_enphase_data, run_forecast_for_enphase
```
these should be imported from the main library

aryanbhosale replied:
> these should be imported from the main library

Yes, will remove these and reuse the existing functions

peterdudfield commented on dashboards/dashboard_2/README.md:
````diff
@@ -9,34 +9,39 @@ https://github.com/user-attachments/assets/940ca4a7-65a4-40ea-a602-ea0f65e96746
-  ENPHASE_API_KEY=your_api_key
-  ENPHASE_SYSTEM_ID=your_system_id
-  ```
+- Check out `.env.example` file and copy paste the contents of the file in the `.env` file that you just created
````
could you add this file too?

aryanbhosale replied:
> could you add this file too?

The env example?

aryanbhosale replied:
> could you add this file too?

Didn't quite understand, could you please elaborate?

peterdudfield replied:
`.env.example` ?

aryanbhosale replied:
> `.env.example` ?

That file's already there in the root directory of the project

peterdudfield replied:
The readme says `Check out `.env.example` file and copy paste the contents of the file in the `.env` file that you just created`. How do i check out the .env.example file?

aryanbhosale replied:
> The readme says `Check out `.env.example`file and copy paste the contents of the file in the`.env` file that you just created`. How do i check out the .env.example file?

The env example file is in the Open-Source-Quartz-Solar-Forecast folder, it's already a part of the main branch of this repo, you can check it out there

peterdudfield commented:
Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR. 

This PR is probably worth merging now, but could we make the follow issue below, and discuss that there

I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard
Then the users can 
1. Enter a site details
2a. Get a forecast with no live data
2b. Get a forecast with live data, by entering credentials for that inverter.

aryanbhosale commented:
> Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR.
> 
> This PR is probably worth merging now, but could we make the follow issue below, and discuss that there
> 
> I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard Then the users can
> 
> 1. Enter a site details
>    2a. Get a forecast with no live data
>    2b. Get a forecast with live data, by entering credentials for that inverter.

Yup that sounds good , it would make it easier for the user to enter their credentials without having to enter them in a separate env file

aryanbhosale commented:
> Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR.
> 
> This PR is probably worth merging now, but could we make the follow issue below, and discuss that there
> 
> I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard Then the users can
> 
> 1. Enter a site details
>    2a. Get a forecast with no live data
>    2b. Get a forecast with live data, by entering credentials for that inverter.

Ill make a video of the working of this

aryanbhosale commented:
> # Pull Request
> ## Description
> created apis for 4 options - no inverter, solis, givenergy, and enphase
> 
> ## Checklist:
> * [x]  My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
> * [x]  I have performed a self-review of my own code
> * [x]  I have made corresponding changes to the documentation
> * [ ]  I have added tests that prove my fix is effective or that my feature works
> * [x]  I have checked my code and corrected any misspellings



> Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR.
> 
> This PR is probably worth merging now, but could we make the follow issue below, and discuss that there
> 
> I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard Then the users can
> 
> 1. Enter a site details
>    2a. Get a forecast with no live data
>    2b. Get a forecast with live data, by entering credentials for that inverter.

ive added a video in the description of the pr

aryanbhosale commented:
> Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR.
> 
> This PR is probably worth merging now, but could we make the follow issue below, and discuss that there
> 
> I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard Then the users can
> 
> 1. Enter a site details
>    2a. Get a forecast with no live data
>    2b. Get a forecast with live data, by entering credentials for that inverter.

1. Enter site details: this is already done in this pr, it gives you an option to choose between the default values and the custom user input values

peterdudfield commented:
> > Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR.
> > This PR is probably worth merging now, but could we make the follow issue below, and discuss that there
> > I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard Then the users can
> > 
> > 1. Enter a site details
> >    2a. Get a forecast with no live data
> >    2b. Get a forecast with live data, by entering credentials for that inverter.
> 
> 1. Enter site details: this is already done in this pr, it gives you an option to choose between the default values and the custom user input values

But they dont enter there credendtials, as they are in the .env file. But if we make a public dashboard then we cant do it this way

aryanbhosale commented:
> > > Ok, I think this is ready to go. @aryanbhosale , would you mind making a video of this dashbaord, and adding it to the PR.
> > > This PR is probably worth merging now, but could we make the follow issue below, and discuss that there
> > > I think we should make a seperate issue, where we want to run a general api and a general dashboard. This could even be a public api / dashboard Then the users can
> > > 
> > > 1. Enter a site details
> > >    2a. Get a forecast with no live data
> > >    2b. Get a forecast with live data, by entering credentials for that inverter.
> > 
> > 
> > 
> > 1. Enter site details: this is already done in this pr, it gives you an option to choose between the default values and the custom user input values
> 
> But they dont enter there credendtials, as they are in the .env file. But if we make a public dashboard then we cant do it this way

Oh that way, yes that's right

peterdudfield commented:
If you fancy putting that video in the dashbaord readme, that would be great?

aryanbhosale commented:
> If you fancy putting that video in the dashbaord readme, that would be great?

will do
